velox/functions/sparksql/ToJson.cpp

```diff
@@ -0,0 +1,540 @@
+
+  AsJson keysAsJson{
+      context, mapKeys, elementsRows, elementToTopLevelRows};
+  AsJson valuesAsJson{
+      context, mapValues, elementsRows, elementToTopLevelRows};
+
+  // Estimates an upperbound of the total length of all Json strings for the
+  // input according to the length of all elements Json strings and the
+  // delimiters to be added.
+  size_t elementsStringSize = 0;
+  context.applyToSelectedNoThrow(rows, [&](auto row) {
+    if (inputMap->isNullAt(row)) {
+      // "null" will be inlined in the StringView.
+      return;
+    }
+
+    auto offset = inputMap->offsetAt(row);
+    auto size = inputMap->sizeAt(row);
+    for (auto i = offset, end = offset + size; i < end; ++i) {
+      // The construction of keysAsJson ensured there is no null in keysAsJson
+      elementsStringSize += keysAsJson.at(i).size() + valuesAsJson.lengthAt(i);
+    }
+
+    // Extra length for commas, semicolons, and curly braces.
+    elementsStringSize += size > 0 ? size * 2 + 1 : 2;
+  });
+
+  flatResult.getBufferWithSpace(elementsStringSize);
+
+  // Constructs the Json string of each map from Json strings of its keys and
+  // values.
+  std::vector<std::pair<StringView, vector_size_t>> sortedKeys;
+  context.applyToSelectedNoThrow(rows, [&](auto row) {
+    if (inputMap->isNullAt(row)) {
+      flatResult.set(row, "null");
+      return;
+    }
+
+    auto offset = inputMap->offsetAt(row);
+    auto size = inputMap->sizeAt(row);
+
+    // Sort entries by keys in each map.
+    sortedKeys.clear();
+    for (int i = offset, end = offset + size; i < end; ++i) {
+      sortedKeys.push_back(std::make_pair(keysAsJson.at(i), i));
+    }
+    std::sort(sortedKeys.begin(), sortedKeys.end());
+
+    auto proxy = exec::StringWriter<>(&flatResult, row);
+
+    proxy.append("{"_sv);
+    for (auto it = sortedKeys.begin(); it != sortedKeys.end(); ++it) {
+      if (it != sortedKeys.begin()) {
+        proxy.append(","_sv);
+      }
+      std::string keyFormat = mapType.childAt(0)->isVarchar() ? "{}:" : "\"{}\":";
+      proxy.append(fmt::format(keyFormat, it->first));
+      valuesAsJson.append(it->second, proxy);
+    }
+    proxy.append("}"_sv);
+
+    proxy.finalize();
+  });
+}
+
+class ToJsonFunction final : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const final {
+    context.ensureWritable(rows, outputType, result);
+    result->clearNulls(rows);
+    auto* rawResults = result->as<FlatVector<StringView>>();
+
+    VELOX_DYNAMIC_TYPE_DISPATCH_ALL(
+        toJson, args[0]->typeKind(), *args[0], context, rows, *rawResults);
+  }
+
+  static std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+    // T(ROW/ARRAY/MAP) -> varchar
+    return {exec::FunctionSignatureBuilder()
+                .typeVariable("T")
```

Comment:
I think we only allow ROW/ARRAY/MAP types as input types, we should add some check

Reply:
Add the input type check in `apply()` method.

Reply:
We should check the type before calling `apply`. Please perform the check in the `initialize` method or the constructor.
Please follow either [this](https://github.com/facebookincubator/velox/blob/5d0e276df398bc7938e21f63d83fa0a7dc93031b/velox/functions/sparksql/LeastGreatest.cpp#L95) or [this](https://github.com/facebookincubator/velox/blob/5d0e276df398bc7938e21f63d83fa0a7dc93031b/velox/functions/prestosql/ArrayFunctions.h#L185).